### PR TITLE
Minimal System.IO.Pipelines integration to prepare for full-async work

### PR DESCRIPTION
**Pull request by stebet**
## Proposed Changes

This introduces `System.IO.Pipelines` into the RabbitMQ Client. I decided to take the minimal step required to add it, as a way to break down PR number #1199 into more manageable parts.

This has no API changes, it still has the Channels for transmission buffering (to be changed later), but instead of reading/writing directly to a `BufferedStream`, we are now using Pipelines which handle the buffering and back-pressure.

This paves the way forward to slowly get rid of the Channels as well, and to create a fully and proper asynchronous code paths for sending/receiving data in a performance manner.

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

- [X] I have read the `CONTRIBUTING.md` document
- [X] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [X] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

### CODE REVIEW

**Comment by lukebakken** [comment]
Thanks, I'll review this coming week.

**Review by lukebakken — APPROVED** [review]
This seems reasonable to me and tests pass, so 👍 . I'm going to wait on another review before merging.

**Review by Zerpet — COMMENTED** [review]
These changes look good to me 👍 Tests pass and both test apps run without errors. Interestingly, `CreateChannel` test app seems to run 1 second faster (7.6s in `main` vs 6.5s) with these changes.

**Review by lukebakken — DISMISSED** [review]

File: projects/RabbitMQ.Client/RabbitMQ.Client.csproj
Comment (danielmarbach) [inline projects/RabbitMQ.Client/RabbitMQ.Client.csproj:64]: Is there a reason not to use 7.x?

File: projects/RabbitMQ.Client/client/impl/Connection.Receive.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Connection.Receive.cs:80]: Nit suggestion: 

```suggestion
                InboundFrame asyncFrame = await _frameHandler.ReadFrameAsync()
                   .ConfigureAwait(false);
```

File: projects/RabbitMQ.Client/client/impl/Connection.Receive.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Connection.Receive.cs:93]: I don't know but something with the just reads weird to me.

```suggestion
                    // Ignore it: we've already recently reset the heartbeat
```

or 

```suggestion
                    // Ignore it: The heartbeat was recently reset
```

File: projects/RabbitMQ.Client/client/impl/Connection.Receive.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Connection.Receive.cs:123]: Nit: 

I would introduce here a local for what is returned by `Lookup`

File: projects/RabbitMQ.Client/client/impl/Connection.Receive.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Connection.Receive.cs:88]: I would prefer `shallReturnPayload`

File: projects/RabbitMQ.Client/client/impl/Connection.Receive.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Connection.Receive.cs:165]:
```suggestion
                        await ClosingLoop()
                           .ConfigureAwait(false);
```

File: projects/RabbitMQ.Client/client/impl/Connection.Receive.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Connection.Receive.cs:188]:
```suggestion
                await ReceiveLoop()
                   .ConfigureAwait(false);
```

File: projects/RabbitMQ.Client/client/impl/Frame.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Frame.cs:231]: I might be missing something. Why is the copy needed here? The buffer should be reachable for the check below and not go out of scope.

File: projects/RabbitMQ.Client/client/impl/Frame.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Frame.cs:239]: Also why would the copy be necessary here? The values should be copied to the ints in the exception anyway shouldn't they? See https://dotnetfiddle.net/XAsUAK

File: projects/RabbitMQ.Client/client/impl/Frame.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Frame.cs:254]: Why is the async suffix used here and not in other async methods?

File: projects/RabbitMQ.Client/client/impl/Frame.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Frame.cs:256]:
```suggestion
            ReadResult result = await reader.ReadAsync()
               .ConfigureAwait(false);
```

File: projects/RabbitMQ.Client/client/impl/Frame.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Frame.cs:261]: Would sort of a throw helper be good here?

File: projects/RabbitMQ.Client/client/impl/Frame.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Frame.cs:271]:
```suggestion
                result = await reader.ReadAsync()
                   .ConfigureAwait(false);
```

File: projects/RabbitMQ.Client/client/impl/Frame.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Frame.cs:270]:
```suggestion
                // Not enough data, read a bit more
```

File: projects/RabbitMQ.Client/client/impl/Frame.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Frame.cs:276]: A throw helper maybe? For example use a static local function?

File: projects/RabbitMQ.Client/client/impl/Frame.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Frame.cs:351]: I know we return without clearing but couldn't we theoretically read garbage when throwing in cases when the returned buffer is already used again by another thread that rented it? Hw about reading the end marker first and then returning?

File: projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs:241]: Async suffix or not?

File: projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs:253]: Would `#if NET` suffice?

File: projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs:254]: Does the stream need to be managed that is returned by `AsStream`?

File: projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs:279]:
```suggestion
                    await _pipeWriter.FlushAsync()
                       .ConfigureAwait(false);
```

File: projects/RabbitMQ.Client/util/NetworkOrderDeserializer.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/util/NetworkOrderDeserializer.cs:54]: Maybe I miss something. Doesn't this under the covers do `MemoryMarshal.Read` which should copy into the ushort? Why is the stackalloc needed?

File: projects/RabbitMQ.Client/util/NetworkOrderDeserializer.cs
Comment (danielmarbach) [inline projects/RabbitMQ.Client/util/NetworkOrderDeserializer.cs:71]: Maybe I miss something. Doesn't this under the covers do `MemoryMarshal.Read` which should copy into the int? Why is the stackalloc needed?

**Review by danielmarbach — COMMENTED** [review]
Slightly late to the party. Sorry!

File: projects/RabbitMQ.Client/client/impl/Connection.Receive.cs
Comment (danielmarbach) [reply projects/RabbitMQ.Client/client/impl/Connection.Receive.cs:80]: You are welcome to ignore. This is just my taste. I think it separates clearly the actual async operation from the `ConfigureAwait` noise.

Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/Connection.Receive.cs:80]: I like the suggestion. I think we'll address this in bulk before 7.0 ships.

File: projects/RabbitMQ.Client/util/NetworkOrderDeserializer.cs
Comment (lukebakken) [reply projects/RabbitMQ.Client/util/NetworkOrderDeserializer.cs:54]: I'm guessing it's due to the fact that `buffer` is a `ReadOnlySequence<byte>` vs what `BinaryPrimitives.ReadUInt16BigEndian' expects (`ReadOnlySpan<byte>`).

I'm not sure why `TryReadUInt16BigEndian` is needed. Since we check that we have enough data to read the `UInt16` value it should always succed.

File: projects/RabbitMQ.Client/client/impl/Frame.cs
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/Frame.cs:239]: Very cool way to demonstrate that!

**Comment by lukebakken** [comment]
Thanks @danielmarbach ... as you can see I'm working through your PR review comments. Thanks a MILLION.

File: projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs:254]: I don't know why `.AsStream()` is necessary here, because `_pipeWriter.Write()` suffices. In addition, no preprocessor directive is necessary.

File: projects/RabbitMQ.Client/client/impl/Frame.cs
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/Frame.cs:256]: I will ensure this is consistent.

Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/Frame.cs:256]: https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/1296

File: projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs:241]: https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/1296

File: projects/RabbitMQ.Client/client/impl/Frame.cs
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/Frame.cs:348]: It seems reasonable to clear the array in this case.

Comment (danielmarbach) [reply projects/RabbitMQ.Client/client/impl/Frame.cs:348]: My comment was not about clearing the array but about the order.

We return `payloadBytes` to the pool which makes it available for use and then throw but access in the exception constructor the end marker with `payloadBytes[payloadSize]` accessing the buffer we have just returned which might already be used by some other thread.

I was asking if the code should look like

```csharp
var endMarker = payloadBytes[payloadSize];
ArrayPool<byte>.Shared.Return(payloadBytes);
throw new MalformedFrameException($"Bad frame end marker: {endMarker}");
```

Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/Frame.cs:348]: OH! I see 👍

File: projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs
Comment (danielmarbach) [reply projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs:254]: I was more wondering if you would need to dispose the stream returned by `AsStream`. It seems though when you dispose it, it does complete the underlying writer

https://github.com/dotnet/runtime/issues/28965
https://github.com/dotnet/corefx/pull/36160

It seems though the PipeStream doesn't necessarily need to be disposed. So maybe adding a comment here would help to make it clear the stream is deliberately not disposed or you could use AsStream and set leaveOpen to true and then dispose the stream when done.

**Comment by stebet** [comment]
Great to see traction here :) thanks @danielmarbach and @lukebakken. I've been less active lately while I'm transituoning into a new job as well as doing other OSS projects and public speaking, but the RabbitMQ client is always on my list of things to revisit :)

File: projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs
Comment (stebet) [reply projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs:254]: It doesn't need to be disposed AFAICT, it mostly provides convenience methods to more easily adapt Stream based APIs to Pipelines.

File: projects/RabbitMQ.Client/util/NetworkOrderDeserializer.cs
Comment (stebet) [reply projects/RabbitMQ.Client/util/NetworkOrderDeserializer.cs:54]: TryRead is faster if there is enough space in the first Span in the Sequence to read the data. If we do not have enough space, we stackalloc enough space (to save heap allocations) and copy the data in there to read without trying.

And yes, the `BinaryPrimitives` methods don't have overloads to read from a `ReadOnlySequence` :(

**Review by Zerpet — DISMISSED** [review]
Looks good to me 👍 Let's give other reviewers some time to comment before we merge.

File: projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs
Comment (inbarbarkai) [reply projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs:254]: Won't the GC eventually call the dispose method causing the writer to be completed?

Comment (danielmarbach) [reply projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs:254]: My input was about making things explicit with a comment or switching to `leaveOpen` to `true` and dispose the stream. Then you never have to ask that question to yourself again when you look at this code a few months down the line

Comment (inbarbarkai) [reply projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs:254]: I agree it is more straightforward, I asked because @lukebakken marked this as resolved.
Made me wonder if I am missing something.

Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs:254]: I can add a comment!

**Comment by stebet** [comment]
> Looks good to me 👍 Let's give other reviewers some time to comment before we merge.

Like your profile pic! I used to work on EVE Online for CCP Games :)

**Review by lukebakken — APPROVED** [review]

**Comment by lukebakken** [comment]
I plan on merging this as soon as CI passes, FYI!

**Comment by lukebakken** [comment]
Thank you @stebet

**Comment by stebet** [comment]
Thanks for picking up the remaining work @lukebakken. Hopefully the full-async work can begin soon :)

**Pull request merged**
